src/frontend/src/optimizer/plan_node/stream_table_scan.rs

```diff
@@ -284,9 +312,21 @@ impl StreamTableScan {
             column_ids: upstream_column_ids.clone(),
         };
 
-        let catalog = self
-            .build_backfill_state_catalog(state)
-            .to_internal_table_prost();
+        let catalog = match self.stream_scan_type {
+            StreamScanType::SnapshotBackfill => self
+                .build_backfill_state_catalog(state, true)
+                .to_internal_table_prost(),
+            StreamScanType::Chain
+            | StreamScanType::Rearrange
+            | StreamScanType::Backfill
+            | StreamScanType::UpstreamOnly
+            | StreamScanType::ArrangementBackfill => self
+                .build_backfill_state_catalog(state, false)
+                .to_internal_table_prost(),
+            StreamScanType::Unspecified => {
+                unreachable!()
+            }
+        };
```

Comment:
Seems cleaner to pass in `StreamScanType` as an argument. Then inside `build_backfill_state_catalog` match on `StreamScanType` and construct the catalog accordingly.